crates/neon/src/thread/mod.rs

```diff
@@ -0,0 +1,232 @@
+//! Thread-local storage for JavaScript threads.
+//!
+//! At runtime, an instance of a Node.js addon can contain its own local storage,
+//! which can then be shared and accessed as needed between Rust modules. This can
+//! be useful for setting up long-lived state that needs to be shared between calls
+//! of an addon's APIs.
+//!
+//! For example, an addon may wish to track the [thread ID][threadId] of each of its
+//! instances:
+//!
+//! ```
+//! # use neon::prelude::*;
+//! # use neon::thread::LocalKey;
+//! static THREAD_ID: LocalKey<u32> = LocalKey::new();
+//!
+//! pub fn thread_id<'cx, C: Context<'cx>>(cx: &mut C) -> NeonResult<u32> {
+//!     THREAD_ID.get_or_try_init(cx, |cx| {
+//!         let global = cx.global();
+//!         let require: Handle<JsFunction> = global.get(cx, "require")?;
+//!         let worker: Handle<JsObject> = require.call_with(cx)
+//!             .arg(cx.string("node:worker_threads"))
+//!             .apply(cx)?;
+//!         let threadId: Handle<JsNumber> = worker.get(cx, "threadId")?;
+//!         Ok(threadId.value(cx) as u32)
+//!     }).cloned()
+//! }
+//! ```
+//!
+//! ### The Addon Lifecycle
+//!
+//! For some use cases, a single shared global constant stored in a `static` variable
+//! might be sufficient:
+//!
+//! ```
+//! static MY_CONSTANT: &'static str = "hello Neon";
+//! ```
+//!
+//! This variable will be allocated when the addon is first loaded into the Node.js
+//! process. This works fine for single-threaded applications, or global immutable
+//! data.
+//!
+//! However, since the addition of [worker threads][workers] in Node v10,
+//! modules can be instantiated multiple times in a single Node process. This means
+//! that while the dynamically-loaded binary library (i.e., the Rust implementation of
+//! the addon) is only loaded once in the running process, but its `main()` function
+//! is executed multiple times with distinct module objects, once per application thread:
+//!
+//! ![The Node.js addon lifecycle, described in detail below.][lifecycle]
+//!
+//! This means that any thread-local data needs to be initialized separately for each
+//! instance of the addon. This module provides a simple container type, [`LocalKey`](LocalKey),
+//! for allocating and initializing thread-local data. (Technically, this data is stored in the
+//! addon's [module instance][environment], which is equivalent to being thread-local.)
+//!
+//! A common example is when an addon defines a custom datatype, which cannot be used across
+//! separate threads. The constructor for the datatype can be [rooted](crate::handle::Root) and
```

Comment:
This might be a bit confusing. It _seems_ to imply it's the *custom* part that makes it unable to be shared. In the example, this can't be shared because it's a reference to a JavaScript value and this is generally true about all `JsValue`.

```suggestion
//! A common example is when an addon needs to maintain a reference to a JavaScript value. References cannot
//! be used across separate threads. The constructor for the datatype can be [rooted](crate::handle::Root) and
```